### PR TITLE
Themes: Isomorphic routing for `/design` [WIP/Experimental]

### PR DESCRIPTION
**Pull request by mcsf**
The idea is to:
- Have the same base code for client and server.
- Once we have full SSR: have just one React factory, in which
  client-side effects like analytics can be encapsulated safely.
- Once we have full SSR: remove server handlers that default to
  renderLoggedInRoute, for more code reuse.
- Take these abstractions further and have less and less code in
  server/pages/index.js. A mechanism similar to that using sections.js
  on the client is one to consider.
- I'm not totally set on the custom 'memo' approach (vs. simple
  memoize), but it's something I'll be exploring some more.
- Probably other things I'm forgetting to mention, and I'm sure you lot
  will ask me about. :)

### TODO

- [ ] Converge to more middleware-based solution, vs. having a separate `selectFactory` concept
- [ ] Fix caching
- [ ] Remove webpack hacks
- [ ] Tests
- [ ] Work with #3053 

The initial commit includes a lot of garbage commented code just so I feel more comfortable deleting it afterwards.

### CODE REVIEW

**Comment by ockham** [comment]
You're introducing `selectFactory` and `selectProps`, which you then wrap in a middleware and add to the middleware stack [here](https://github.com/Automattic/wp-calypso/pull/3065/files#diff-fed100f36ed254ecf34220dcd14f50d8R19). I was wondering if we could omit those extra entities and just define that middleware explicitly in the individual section's `controller.js`, and link them to middlewares in `index.js`, much like we're doing now (and even more so since it still all ends up in a router definition linking a route to a bunch of middlewares anyway). Side effects like analytics handling could go to a different middleware, which is only called on the client. Main advantages would be

1. fewer entities to wrap our heads around and maintain
2. structures that others are familiar with already (routing via middleware), aiding adoption (instead of an additional new structure)

I'm thinking of something (roughly) like

```js
export function routing( router ) {
	if ( isLoggedOut ) {
		route( '/design', themesController.loggedOut );
	} else {
		route( '/design/*', controller.navigation, controller.siteSelection );
		route( '/design', themesController.multiSite );
		route( '/design/:site_id', themesController.singleSite );		
	}
	route( '/design/type/:tier', themesController.multiSite );
	route( '/design/type/:tier/:site_id', themesController.singleSite )

}

```

(possibly with some other conditionals for client-side only etc). This could then be called with `router` set to `page` from `sections`, and with `router` set to `app.get` on the server. (I'd need to check if express can also [separate concerns like `page.js`](https://github.com/visionmedia/page.js#separating-concerns)). But that same things could certainly also be done more functionally :-)

**Comment by ehg** [comment]
> structures that others are familiar with already (routing via middleware), aiding adoption (instead of an additional new structure)

I agree with this :)

File: webpack.config.node.js [hunk 1]
```diff
@@ -73,7 +73,33 @@ module.exports = {
 	},
 	plugins: [
 		// Require source-map-support at the top, so we get source maps for the bundle
-		new webpack.BannerPlugin( 'require( "source-map-support" ).install();', { raw: true, entryOnly: false } )
+		new webpack.BannerPlugin( 'require( "source-map-support" ).install();', { raw: true, entryOnly: false } ),
+		new webpack.NormalModuleReplacementPlugin( /^my-sites\/controller$/, 'lodash/utility/noop' ),
```
Comment (ehg) [inline webpack.config.node.js]: We need a better solution for this. Perhaps we could include isomorphic and non-isomorphic middlewares/controllers using a package.json approach. We'd eventually get rid of any non-isomorphic stuff there.

**Comment by mcsf** [comment]
> structures that others are familiar with already (routing via middleware), aiding adoption (instead of an additional new structure)

I also agree. Also, your comment reveals that you've looked at this deeply, so thanks. :)

`selectFactory` was the product of my dealing with complexity by thinking outside of the current architecture's premises. As the pieces fall into place and the WIP actually starts working, it gets easier to know what to refine.

> I'm thinking of something (roughly) like
> ```js
> export function routing( router ) {
> ```

Your example with a `router()` helper is nice. While this judgment is impaired by some sleep deprivation and recent stress :), I have my doubts as to how easily it can be implemented on the server, though. Mainly because of a crucial difference between client and server: on the client, `isLoggedOut` is a constant throughout the app's lifetime, so we can setup routing statically, so to speak. We don't have that on the server, though, so — to stick with a `router()`-like approach — we'd have either 1) entirely dynamic routing (i.e. one endpoint for all Themes paths, which would assess what to do next), or 2) hybrid routing with static routes for all paths, but each route handler would have some logic injected to assess whether to treat the request as logged in or out. The function pipelining achieved with `select*` and friends is a way to address this.

(^ I hope this made any sense.)

Notwithstanding, I'm all for reusing what we have and making the most of middleware.

**Comment by mcsf** [comment]
How can we recycle existing Calypso concepts? We have "middleware" (route handlers w/ continuation via `next()`); can't we just use that? Right now, on the client, we have:
- 2 optional (logged-in-only) middleware/controllers (`navigation`, `siteSelection`)
- 1 catch-all middleware that calls `selectFactory`, `selectProps`, and renders

A separate catch-all handler is used on the server, that reuses `select*` and adds caching.

**What if** we could have everything defined in terms of middleware that may have effects and may populate (read: mutate) the `context` object of that request/route change? Along the middleware chain, `context` could be filled with the necessary data for rendering. For instance:

```js
page( 'cool-path', navigation, siteSelection, prepareLayout, render )
```

_Example middleware and its effects_:
- [client-only] **`controller.navigation` & `controller.siteSelection`**: (set up nav and site sel)
- [iso] **`prepareLayout`**: `context.layoutFactory = …; context.props = …`
- [server-only] **`maybeCache`**: `if ( serverSide && cacheHit ) context.layout = …`
- [iso or not] **`render`**: `if ( ! context.layout ) context.layout = …` (can be markup or nodes)

```js
const makeServerRouteHandler = ( … ) => ( req, res ) => {
  const ctx = getEnhancedContext( req )
  const mws = getMiddleware( ctx )
  applyMiddleware( mws )( ctx )
  res.render( 'index.jade', context ) // or: abstracted into middleware chain if we want
}

// this would be the equivalent to the current `getRouting`, filtering
// for the current conditions (e.g. server-side? logged in?)
const getMiddleware = ( ctx ) => {
  // magic
  return filterMatching( _middleware, ctx )
}

const makeClientRouteHandler = ( … ) => ( ctx ) => {
  applyMiddleware( getMiddleware( ctx ) )( ctx )
  render( ctx.layout, domNode ) // or abstracted, or other rendering method
}

// example middleware
const prepareLayout = ( ctx, next ) => {
  let matches
  let layoutFactory = ThemesMain
  let props = {}
  const { path } = ctx
  // use `path-parser` or something
  const p = new Path( path )

  if ( matches = p.match( '/design/type/:tier' ) ) {
    props = { ...props, matches }
  }

  // etc.

  // then, unless something went wrong,
  Object.assign( ctx, { layoutFactory, props } )
  next()
}
```

I felt like trying this on paper first (headache-induced screen avoidance). I'd like to hear thoughts on this.

**Comment by mcsf** [comment]
Sharing the gist that @ockham and I voice-paired on while we let the ideas stir:
https://gist.github.com/mcsf/c8006dc383166f284f6a

**Comment by mcsf** [comment]
This was a nice prototyping experience, but closing this in favor of other isomorphic routing PRs, like #3207.

**Pull request closed without merging**